[Swa Yong Shen] iP

swayongshen opened the pull request.

Commit:
Added level 1

Commit:
Level-2

Commit:
Level-3

Commit:
Level 4

Commit:
Added exception class DukeException:

Changed handleInput to throw DukeException.
Changed addToDo to throw DukeException.

Commit:
addDeadLine throws DukeException

Commit:
doneTask throws DukeException

Commit:
addEvent throws DukeException

Commit:
Level-5

Commit:
Added deleteTask

Commit:
Test

Commit:
Removed Test

swayongshen committed:
Testing git username

swayongshen committed:
Refactor printing output to Printer class

Printing output should not be the job of Duke, which is to handle input.
As per single responsibility principle, move printing functions to a
separate class.

swayongshen committed:
Add auto testing

swayongshen committed:
Added Javadoc comments to relevant methods

swayongshen committed:
Restore point for Level-7

swayongshen committed:
Added read and write task to file

swayongshen committed:
Finish read and write from file prototype

swayongshen committed:
Shift REGEX constants into stringToTask

swayongshen committed:
Complete minimal goal of Level-8

swayongshen committed:
Fix done task to also write changes to file

swayongshen committed:
Merge branch level 8

swayongshen committed:
Add class Storage and class Ui, refactor methods

swayongshen committed:
Add Command classes

swayongshen committed:
Add classes DeleteCommand, AddCommand

swayongshen committed:
Finish abstracting Parser, Command

swayongshen committed:
Several changes

* Increase number of date formats accepted for deadline and event.
* Refactor stringToTask to be in Parser instead of Task.
* Fix date parsing error when parsing tasks from the storage.

swayongshen committed:
Added stretch goals for Level 8

swayongshen committed:
Add level-8 stretch goal features

swayongshen committed:
Merge branch 'master' of https://github.com/swayongshen/ip into master

swayongshen committed:
Compartmentalize files into packages

swayongshen committed:
Fix some Duke that Intellij changed to duke.Duke

swayongshen committed:
Add JUnit tests for DatedTask and Parser classes

swayongshen committed:
Add JavaDoc to most public methods

swayongshen committed:
Add FindCommand class to handle find commands

swayongshen committed:
Add Blank file to enable merge commit

swayongshen committed:
Merge branch 'branch-A-CodingStandard'

swayongshen committed:
Merge branch-Level-9 to master

swayongshen committed:
Fix file structure of JUnit tests

swayongshen committed:
Add checkstyle file and corrected style errors

swayongshen committed:
Add Gradle support, remove CS2030's checkstyle, fix checkstyle errors

swayongshen committed:
Refactor Duke and relevant classes' methods.

Refactor methods to pass String as return value around.

Reason: ui.MainWindow requires a string to be passed in via
Duke.getResponse method so that it can display Duke's response to ther
user via the GUI.

swayongshen committed:
Delete duke.Ui class, move Ui.formatStringArray to duke.Helper

swayongshen committed:
Add duke.Main class to bridge Duke and duke.ui.MainWindow

swayongshen committed:
Latest update for GUI

swayongshen committed:
Add GUI files

swayongshen committed:
Make GUI Work

swayongshen committed:
Refactor JavaFX related classes to duke.ui package

kieron560 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,37 @@
+
+    /**
+     *  Provides responses to commands that are passed in by the user.
+     */
```
Should you have @params and @returns here?

kieron560 commented on src/main/java/duke/Helper.java:
```diff
@@ -0,0 +1,64 @@
+     * @param end end index of sub-array to join
+     * @return string of sub-array joined with space
+     */
+    public static String join(String[] arr, int start, int end) {
```
I think there should be punctuations behind the parameter description? As per the Java Coding Standards.

kieron560 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,94 @@
+        String task = Helper.join(userInputSplit, 1, userInputSplit.length - 1);
+        return list.add(new ToDo(task));
+    }
+
```
Should you add a JavaDocs here?

kieron560 reviewed (commented):
Overall looks good, perhaps javadocs changes!

zhoube commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,113 @@
+        this.list = storage.readTasksFromFile();
+    }
+
+    /**
```
Should there be a @return and @throws comment here?

zhoube commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    /**
+     * Marks a task as done and also rewrite all tasks in the storage to reflect this change.
```
Should there be a @return here as well?

zhoube reviewed (commented):
Very well done! I believe you have adhered to almost all of the Java Coding Standards, perhaps you could work on some of the JavaDoc comments that were missing here!

swayongshen committed:
Change Ui.formatStrings to use varargs as argument

swayongshen committed:
Add and modify Javdoc comments

swayongshen committed:
Add assertions on relevant methods

swayongshen committed:
Check code quality of duke.command package

swayongshen committed:
Check code quality of duke.task package

swayongshen committed:
Check code quality of duke package

swayongshen committed:
Merge pull request #4 from swayongshen/branch-A-CodeQuality

Improve code quality.

swayongshen committed:
Merge master changes

swayongshen committed:
Merge pull request #3 from swayongshen/branch-A-Assertions

Add assertions on relevant methods

swayongshen committed:
Add code quality checklist text file

swayongshen committed:
Merge branch 'master' of github.com:swayongshen/ip

swayongshen committed:
Add CI for Github

swayongshen committed:
Move gradle.yml to correct directory

swayongshen committed:
Add Javadoc comments, fix code style issues

swayongshen committed:
Fix TaskList.remove()

Fix referencing a task in the task list after the task was deleted
because it led to array access error. Instead, move the reference
forward.

swayongshen committed:
Remove Jacoco from Gradle

swayongshen committed:
Add multi delete and sort by date/task

cheunggalen commented on src/main/java/duke/Helper.java:
```diff
@@ -0,0 +1,121 @@
+            return false;
+        }
+        for (int i = 0; i < s.length(); i++) {
+            if (i == 0 && s.charAt(i) == '-') {
+                if (s.length() == 1) {
+                    return false;
+                } else {
+                    continue;
+                }
+            }
```
I'm not sure if this is considered somewhat of a deep nesting, resulting in a arrow head structure?

cheunggalen commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,122 @@
+     * @param userInput user command input to be parsed
+     * @throws DukeException if command is in an incorrect format.
+     */
+    public static Command handleInput(String userInput) throws DukeException {
+        String[] splitBySpaces = userInput.trim().split("\\s+");
+        String keyword = splitBySpaces[0];
+        if (keyword.equals("list")) {
+            return new ListCommand(splitBySpaces);
+        } else if (keyword.equals("done")) {
+            return new DoneCommand(splitBySpaces);
+        } else if (keyword.equals("deadline") || keyword.equals("todo") || keyword.equals("event")) {
+            return new AddCommand(splitBySpaces);
+        } else if (keyword.equals("delete")) {
+            return new DeleteCommand(splitBySpaces);
+        } else if (keyword.equals("bye")) {
+            return new ByeCommand(splitBySpaces);
+        } else if (keyword.equals("find")) {
+            return new FindCommand(splitBySpaces);
+        } else if (keyword.equals("sort")) {
+            return new SortCommand(splitBySpaces);
+        } else {
+            throw new DukeException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
+        }
+    }
```
I like how there is no deep nesting (arrowhead) here!

cheunggalen commented on src/main/java/duke/command/SortCommand.java:
```diff
@@ -0,0 +1,57 @@
+        //Ascending or descending
+        String sortDirection = commandSplit[2];
+
+        if (sortParameter.equals("date")) {
+
+            if (sortDirection.equals("asc") || sortDirection.equals("desc")) {
+                boolean isAscending = sortDirection.equals("asc");
+                return list.sortByDate(isAscending);
+            }
+
+            //Neither asc nor desc
+            throw new DukeException("Please enter a valid sort direction \"asc\" or \"desc\".");
+        }
+
+        if (sortParameter.equals("task")) {
+
+            if (sortDirection.equals("asc") || sortDirection.equals("desc")) {
+                boolean isAscending = sortDirection.equals("asc");
+                return list.sortByDescription(isAscending);
+            }
+
+            //Neither asc nor desc
+            throw new DukeException("Please enter a valid sort direction \"asc\" or \"desc\".");
+        }
+
+        //Neither date nor task
+        throw new DukeException("Please enter a valid sort parameter \"date\" or \"task\".");
```
I like how you made the happy path prominent here!

cheunggalen commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,45 @@
+        return Helper.formatStrings("Nice! I've marked this task as done:", this.toString());
+    }
+
+    public boolean isDescriptionContainsString(String input) {
```
I like how you your method name accurately explains to the reader of the method's functionality. Even if it may be long, it is clear.

cheunggalen commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,60 @@
+            String enteredDate = Helper.join(this.commandSplit, 1, this.commandSplit.length - 1);
+            LocalDate queryDate = parseDate(enteredDate);
+            //Filter Tasks from TaskList against the date provided
+            ArrayList<Task> matchedTasks = list.filter(x -> !x.getDate().equals(LocalDate.MIN)
+                    && x.getDate().equals(queryDate));
```
Could this be done in steps? I'm not sure if I like this complicated expression as it makes it harder to read.

cheunggalen commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,34 @@
+     * @param list the task list.
+     */
+    @Override
+    public String execute(TaskList list) {
+        String keyword = Helper.join(this.commandSplit, 1);
+        ArrayList<Task> filteredList = list.filter(task -> task.isDescriptionContainsString(keyword));
+        List<String> tasksAsString = filteredList.stream().map(Task::toString).collect(Collectors.toList());
+        tasksAsString.add(0, "Found " + tasksAsString.size() + " matching task(s):");
+        return Helper.formatStrings(tasksAsString.toArray(new String[0]));
+    }
```
Could this be abstracted further to follow the SLAP?

cheunggalen commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,34 @@
+     * Instantiates a new DoneCommand object.
+     * @param commandSplit user command split by spaces.
+     */
+    public DoneCommand(String[] commandSplit) {
+        super(commandSplit);
+        assert commandSplit.length > 1 && commandSplit[0].equals("done")
+                && commandSplit[1].chars().allMatch(Character::isDigit)
+                : "Must have done keyword and integer as argument";
+    }
```
I like how you split a long line into several short ones to enhance code readability.

cheunggalen reviewed (commented):
LGTM, almost. Just a few nits (as pointed by me and other reviewers) to fix. Overall, job well done!

swayongshen committed:
Add automatic greeting message

swayongshen committed:
Add static exit function for Main

swayongshen committed:
Change ByeCommand to terminate program.

1. Change Command.execute to return Pair<String, CallbackFunction> instead so that
a CallbackFunction object can be run by MainWindow.java after displaying
the String response from the commands.

1. ByeCommand now passes a CallbackFunction as part of its return pair
which terminates the program when run.

swayongshen committed:
Add representative screenshot

swayongshen committed:
Add user guide

swayongshen committed:
Set theme jekyll-theme-minimal

swayongshen committed:
Delete _config.yml

Remove jekyll theme.

swayongshen committed:
Update README.md

Fix headers of commands in usage section.

swayongshen committed:
Fix checkstyle errors

swayongshen committed:
Fix file not found error when storage not established

swayongshen committed:
Add date format note to user guide

swayongshen committed:
Change delete command usage in user guide.

Reflect the fact that multiple task numbers can be provided to the
delete command.

swayongshen committed:
Change find command to be case insensitive

swayongshen committed:
Add find and sort command to feature list

swayongshen committed:
Fix checkstyle error

swayongshen committed:
Fix GUI label showing ... if text is too long

swayongshen committed:
Change list command to show empty message if list is empty

swayongshen committed:
Remove emoticon from error message

swayongshen committed:
Change name of jar archiveName in gradle

swayongshen committed:
Finishing touches
